src/utils/rw_spin_lock.hpp

```diff
@@ -0,0 +1,114 @@
+// Copyright 2023 Memgraph Ltd.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt; by using this file, you agree to be bound by the terms of the Business Source
+// License, and you may not use this file except in compliance with the Business Source License.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+#pragma once
+
+namespace memgraph::utils {
+namespace {
+/// A helper for RWSpinLock, allows a contended spin lock to yield to another thread.
+struct yeilder {
+  void operator()() noexcept {
+#if defined(__i386__) || defined(__x86_64__)
+    // TODO: make portable
+    __builtin_ia32_pause();
+#endif
+    ++count;
+    if (count > 8) [[unlikely]] {
+      count = 0;
+      nanosleep(&shortpause, nullptr);
+      // Increase the backoff
+      shortpause.tv_nsec = std::min<decltype(shortpause.tv_nsec)>(shortpause.tv_nsec << 1, 512);
+    }
+  }
+
+ private:
+  uint_fast32_t count{0};
+  timespec shortpause = {.tv_sec = 0, .tv_nsec = 1};
+};
+}  // namespace
+
+/**
+ * A reader/writer spin lock.
+ * Stores in a uint32_t,
+ * 0x0000'0001 - is the write bit
+ * rest of the bits hold the count for the number of current writers.
+ * The lock is friendly to writers.
+ * - writer lock() will wait for all readers to leave unlock_shared()
+ * - new reader lock_shared() will wait until writer has unlock()
+ **/
+struct RWSpinLock {
+  RWSpinLock() = default;
+
+  void lock() {
+    // spin: to grant the UNIQUE_LOCKED bit
+    while (true) {
+      // optimistic: assume we will be granted the lock
+      auto const phase1 = std::atomic_ref{lockStatus_}.fetch_or(UNIQUE_LOCKED, std::memory_order_acq_rel);
```

Comment:
can we be explicit about the type of phase1 and phase2, I find it much easier to read the code?